### ISSUE TITLE
AmazonS3Client.GetObject with key null or empty returns an XML with result of ListObjects

### ISSUE THREAD

**Issue by kobi** [aws/aws-sdk-net#266]
When calling `AmazonS3Client.GetObject ` with `path` `null` or `""`, the result is an XML with the content of `ListObjects` - a list of 1000 files in the bucket:

```csharp
public async Task<string> GetContentAsString(string path)
{
	using (var s3 = new AmazonS3Client())
	{
		var response = await s3.GetObjectAsync(BucketName, path);
		using (var reader = new StreamReader(response.ResponseStream))
		{
			var content = await reader.ReadToEndAsync();
			return content;
		}
	}
}
```

I saw the same behavior in all overloads of `ListObjects` and `ListObjectsAsync`. I didn't check if [`GET Object`](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html) behaves the same when invoked directly using the REST API, but either way I think this behavior is unexpected (and not useful), and an exception would have been better.

(yes, we forgot to check for `null` and allowed our customers to download that file. embarrassing.)

**Comment by normj** [comment]
Normally we don't do client side validation in the SDK. This is to make sure the SDK doesn't get stale and force users to upgrade if the service makes changes to the requirements for parameter values.

In this case I do agree with you we should add the validation since not setting the key property transforms the operation into a completely different operation. Which can cause unpredictable behavior in application code.

I'll post back when I make the change.

**Comment by normj** [comment]
We released version 3.1.3.5 of the S3 NuGet package which contains the client side validation to avoid this from happening.
